README.md
# OctoPrint-Domoticz

This plugin is to control switches via Domoticz server.

[Changelog](changelog.md)

##  Screenshots
![screenshot](screenshot.png)

![screenshot](settings.png)

![screenshot](domoticz_editor.png)

## Setup

Install via the bundled [Plugin Manager](https://github.com/foosel/OctoPrint/wiki/Plugin:-Plugin-Manager)
or manually using this URL:

    https://github.com/jneilliii/OctoPrint-Domoticz/archive/master.zip


## Configuration

Once installed go into settings and enter the details for your Domoticz server connection and switch index. Adjust additional settings as needed.

## Settings Explained

- **IP:PORT**
  - The ip and port of Domoticz server.
- **Index**
  - Index number reprensenting the switch to control.
- **Icon**
  - Icon class name from the [fontawesome](https://fontawesome.com/v3.2.1/icons/) library.
- **Label**
  - Title attribute on icon that shows on mouseover.
- **Username**
  - Username used to connect to web interface. If authentication is not configured on your Domoticz server leave this blank.
- **Password**
  - Password used to connect to web interface. If authentication is not configured on your Domoticz server leave this blank.
- **Warn**
  - The left checkbox will always warn when checked.
  - The right checkbox will only warn when printer is printing.
- **GCODE**
  - When checked this will enable the processing of M80 and M81 commands from gcode to power on/off plug.  Syntax for gcode command is M80/M81 followed by hostname/ip and index.  For example if your plug is 192.168.1.2:8080 and index of 1 your gcode command would be **M80 192.168.1.2:8080 1**
  - As of version 0.0.3 you can now use the custom gcode commands `@DOMOTICZON` and `@DOMOTICZOFF`. The difference here is that you only need to supply the index value with the command.  For example `@DOMOTICZOFF 1` will turn the switch with an Index of 1 off.
- **postConnect**
  - Automatically connect to printer after plug is powered on.
  - Will wait for number of seconds configured in **Auto Connect Delay** setting prior to attempting connection to printer.
- **preDisconnect**
  - Automatically disconnect printer prior to powering off the plug.
  - Will wait for number of seconds configured in **Auto Disconnect Delay** prior to powering off the plug.
- **Cmd On**
  - When checked will run system command configured in **System Command On** setting after a delay in seconds configured in **System Command On Delay**.
- **Cmd Off**
  - When checked will run system command configured in **System Command Off** setting after a delay in seconds configured in **System Command Off Delay**.
  
## Support My Efforts
I programmed this plugin for fun and do my best effort to support those that have issues with it, please return the favor and support me.

[![paypal](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif)](https://paypal.me/jneilliii)




changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.1.2] - 2020-03-27
### Fixed
- Moved system commands and printer connection logic to server-side processing.

## [0.1.1] - 2019-10-22
### Fixed
- Issues introduced in last release that prevented the plugin from working.

## [0.1.0] - 2019-10-06
### Added
- Python 3 compatibility

## [0.0.3] - 2018-11-03
### Added
- Added custom `@DOMOTICZON` and `@DOMOTICZOFF` GCODE commands.
- Bypass previously delayed power off trigger if `Warn While Printing` option is enabled.

## [0.0.2] - 2018-07-07
### Added
- User authentication to Domoticz server.

### Notes
- If authenthication is not enabled leave username and password blank.
- Passwords are stored in plain text and encoded during web call.

## [0.0.1] - 2018-07-06
### Added
- Initial release.

[0.1.1]: https://github.com/jneilliii/OctoPrint-Domoticz/tree/0.1.1
[0.1.0]: https://github.com/jneilliii/OctoPrint-Domoticz/tree/0.1.0
[0.0.3]: https://github.com/jneilliii/OctoPrint-Domoticz/tree/0.0.3
[0.0.2]: https://github.com/jneilliii/OctoPrint-Domoticz/tree/0.0.2
[0.0.1]: https://github.com/jneilliii/OctoPrint-Domoticz/tree/0.0.1


octoprint_domoticz/static/js/domoticz.js
/*
 * View model for OctoPrint-Domoticz
 *
 * Author: jneilliii
 * License: AGPLv3
 */
$(function() {
	function domoticzViewModel(parameters) {
		var self = this;

		self.settings = parameters[0];
		self.loginState = parameters[1];

		self.arrSmartplugs = ko.observableArray();
		self.isPrinting = ko.observable(false);
		self.gcodeOnString = function(data){return 'M80 '+data.ip()+' '+data.idx();};
		self.gcodeOffString = function(data){return 'M81 '+data.ip()+' '+data.idx();};
		self.selectedPlug = ko.observable();
		self.processing = ko.observableArray([]);

		self.onBeforeBinding = function() {
			self.arrSmartplugs(self.settings.settings.plugins.domoticz.arrSmartplugs());
		}

		self.onAfterBinding = function() {
			self.checkStatuses();
		}

		self.onEventSettingsUpdated = function(payload) {
			self.settings.requestData();
			self.arrSmartplugs(self.settings.settings.plugins.domoticz.arrSmartplugs());
		}

		self.onEventPrinterStateChanged = function(payload) {
			if (payload.state_id == "PRINTING" || payload.state_id == "PAUSED"){
				self.isPrinting(true);
			} else {
				self.isPrinting(false);
			}
		}

		self.addPlug = function() {
			self.selectedPlug({'ip':ko.observable(''),
							   'idx':ko.observable('1'),
							   'displayWarning':ko.observable(true),
							   'warnPrinting':ko.observable(false),
							   'gcodeEnabled':ko.observable(false),
							   'gcodeOnDelay':ko.observable(0),
							   'gcodeOffDelay':ko.observable(0),
							   'autoConnect':ko.observable(true),
							   'autoConnectDelay':ko.observable(10.0),
							   'autoDisconnect':ko.observable(true),
							   'autoDisconnectDelay':ko.observable(0),
							   'sysCmdOn':ko.observable(false),
							   'sysRunCmdOn':ko.observable(''),
							   'sysCmdOnDelay':ko.observable(0),
							   'sysCmdOff':ko.observable(false),
							   'sysRunCmdOff':ko.observable(''),
							   'sysCmdOffDelay':ko.observable(0),
							   'currentState':ko.observable('unknown'),
							   'btnColor':ko.observable('#808080'),
							   'username':ko.observable(''),
							   'password':ko.observable(''),
							   'icon':ko.observable('icon-bolt'),
							   'label':ko.observable('')});
			self.settings.settings.plugins.domoticz.arrSmartplugs.push(self.selectedPlug());
			$("#DomoticzEditor").modal("show");
		}

		self.editPlug = function(data) {
			self.selectedPlug(data);
			$("#DomoticzEditor").modal("show");
		}

		self.removePlug = function(row) {
			self.settings.settings.plugins.domoticz.arrSmartplugs.remove(row);
		}

		self.cancelClick = function(data) {
			self.processing.remove(data.ip());
		}

		self.onDataUpdaterPluginMessage = function(plugin, data) {
			if (plugin != "domoticz") {
				return;
			}

			plug = ko.utils.arrayFirst(self.settings.settings.plugins.domoticz.arrSmartplugs(),function(item){
				return ((item.ip().toUpperCase() == data.ip.toUpperCase()) && (item.idx() == data.idx));
				}) || {'ip':data.ip,'idx':data.idx,'currentState':'unknown','btnColor':'#808080','gcodeEnabled':false};

			if(self.settings.settings.plugins.domoticz.debug_logging()){
				console.log(self.settings.settings.plugins.domoticz.arrSmartplugs());
				console.log('msg received:'+JSON.stringify(data));
				console.log('plug data:'+ko.toJSON(plug));
			}

			if (plug.currentState != data.currentState) {
				plug.currentState(data.currentState)
				switch(data.currentState) {
					case "on":
						break;
					case "off":
						break;
					default:
						new PNotify({
							title: 'Domoticz Error',
							text: 'Status ' + plug.currentState() + ' for ' + plug.ip() + '. Double check IP Address\\Hostname in Domoticz Settings.',
							type: 'error',
							hide: true
							});
				}
				self.settings.saveData();
			}
			self.processing.remove(data.ip);
		};

		self.toggleRelay = function(data) {
			self.processing.push(data.ip());
			switch(data.currentState()){
				case "on":
					self.turnOff(data);
					break;
				case "off":
					self.turnOn(data);
					break;
				default:
					self.checkStatus(data);
			}
		}

		self.turnOn = function(data) {
			self.sendTurnOn(data);
		}

		self.sendTurnOn = function(data) {
			$.ajax({
				url: API_BASEURL + "plugin/domoticz",
				type: "POST",
				dataType: "json",
				data: JSON.stringify({
					command: "turnOn",
					ip: data.ip(),
					idx: data.idx(),
					username: data.username(),
					password: data.password()
				}),
				contentType: "application/json; charset=UTF-8"
			});
		};

		self.turnOff = function(data) {
			if((data.displayWarning() || (self.isPrinting() && data.warnPrinting())) && !$("#DomoticzWarning").is(':visible')){
				self.selectedPlug(data);
				$("#DomoticzWarning").modal("show");
			} else {
				$("#DomoticzWarning").modal("hide");
				self.sendTurnOff(data);
			}
		}; 

		self.sendTurnOff = function(data) {
			$.ajax({
			url: API_BASEURL + "plugin/domoticz",
			type: "POST",
			dataType: "json",
			data: JSON.stringify({
				command: "turnOff",
					ip: data.ip(),
					idx: data.idx(),
					username: data.username(),
					password: data.password()
			}),
			contentType: "application/json; charset=UTF-8"
			});
		}

		self.checkStatus = function(data) {
			$.ajax({
				url: API_BASEURL + "plugin/domoticz",
				type: "POST",
				dataType: "json",
				data: JSON.stringify({
					command: "checkStatus",
					ip: data.ip(),
					idx: data.idx(),
					username: data.username(),
					password: data.password()
				}),
				contentType: "application/json; charset=UTF-8"
			}).done(function(){
				self.settings.saveData();
				});
		}; 

		self.disconnectPrinter = function() {
			$.ajax({
				url: API_BASEURL + "plugin/domoticz",
				type: "POST",
				dataType: "json",
				data: JSON.stringify({
					command: "disconnectPrinter"
				}),
				contentType: "application/json; charset=UTF-8"
			});
		}

		self.connectPrinter = function() {
			$.ajax({
				url: API_BASEURL + "plugin/domoticz",
				type: "POST",
				dataType: "json",
				data: JSON.stringify({
					command: "connectPrinter"
				}),
				contentType: "application/json; charset=UTF-8"
			});
		}

		self.sysCommand = function(sysCmd) {
			$.ajax({
				url: API_BASEURL + "plugin/domoticz",
				type: "POST",
				dataType: "json",
				data: JSON.stringify({
					command: "sysCommand",
					cmd: sysCmd
				}),
				contentType: "application/json; charset=UTF-8"
			});
		}

		self.checkStatuses = function() {
			ko.utils.arrayForEach(self.settings.settings.plugins.domoticz.arrSmartplugs(),function(item){
				if(item.ip() !== "") {
					if(self.settings.settings.plugins.domoticz.debug_logging()){
						console.log("checking " + item.ip() + " index " + item.idx());
					}
					self.checkStatus(item);
				}
			});
		};
	}

	// view model class, parameters for constructor, container to bind to
	OCTOPRINT_VIEWMODELS.push([
		domoticzViewModel,

		// e.g. loginStateViewModel, settingsViewModel, ...
		["settingsViewModel","loginStateViewModel"],

		// "#navbar_plugin_domoticz","#settings_plugin_domoticz"
		["#navbar_plugin_domoticz","#settings_plugin_domoticz"]
	]);
});


octoprint_domoticz/__init__.py
# coding=utf-8
from __future__ import absolute_import

import octoprint.plugin
from octoprint.server import user_permission
import socket
import time
import logging
import os
import re
import requests
import threading
import base64

class domoticzPlugin(octoprint.plugin.SettingsPlugin,
                            octoprint.plugin.AssetPlugin,
                            octoprint.plugin.TemplatePlugin,
							octoprint.plugin.SimpleApiPlugin,
							octoprint.plugin.StartupPlugin):

	def __init__(self):
		self._logger = logging.getLogger("octoprint.plugins.domoticz")
		self._domoticz_logger = logging.getLogger("octoprint.plugins.domoticz.debug")

	##~~ StartupPlugin mixin

	def on_startup(self, host, port):
		# setup customized logger
		from octoprint.logging.handlers import CleaningTimedRotatingFileHandler
		domoticz_logging_handler = CleaningTimedRotatingFileHandler(self._settings.get_plugin_logfile_path(postfix="debug"), when="D", backupCount=3)
		domoticz_logging_handler.setFormatter(logging.Formatter("[%(asctime)s] %(levelname)s: %(message)s"))
		domoticz_logging_handler.setLevel(logging.DEBUG)

		self._domoticz_logger.addHandler(domoticz_logging_handler)
		self._domoticz_logger.setLevel(logging.DEBUG if self._settings.get_boolean(["debug_logging"]) else logging.INFO)
		self._domoticz_logger.propagate = False

	def on_after_startup(self):
		self._logger.info("Domoticz loaded!")

	##~~ SettingsPlugin mixin

	def get_settings_defaults(self):
		return dict(
			debug_logging = False,
			arrSmartplugs = [{'ip':'','displayWarning':True,'idx':'1','warnPrinting':False,'gcodeEnabled':False,'gcodeOnDelay':0,'gcodeOffDelay':0,'autoConnect':True,'autoConnectDelay':10.0,'autoDisconnect':True,'autoDisconnectDelay':0,'sysCmdOn':False,'sysRunCmdOn':'','sysCmdOnDelay':0,'sysCmdOff':False,'sysRunCmdOff':'','sysCmdOffDelay':0,'currentState':'unknown','btnColor':'#808080','username':'','password':'','icon':'icon-bolt','label':''}],
		)

	def on_settings_save(self, data):
		old_debug_logging = self._settings.get_boolean(["debug_logging"])

		octoprint.plugin.SettingsPlugin.on_settings_save(self, data)

		new_debug_logging = self._settings.get_boolean(["debug_logging"])
		if old_debug_logging != new_debug_logging:
			if new_debug_logging:
				self._domoticz_logger.setLevel(logging.DEBUG)
			else:
				self._domoticz_logger.setLevel(logging.INFO)

	def get_settings_version(self):
		return 3

	def on_settings_migrate(self, target, current=None):
		if current is None or current < self.get_settings_version():
			# Reset plug settings to defaults.
			self._logger.debug("Resetting arrSmartplugs for domoticz settings.")
			self._settings.set(['arrSmartplugs'], self.get_settings_defaults()["arrSmartplugs"])

	##~~ AssetPlugin mixin

	def get_assets(self):
		return dict(
			js=["js/domoticz.js"],
			css=["css/domoticz.css"]
		)

	##~~ TemplatePlugin mixin

	def get_template_configs(self):
		return [
			dict(type="navbar", custom_bindings=True),
			dict(type="settings", custom_bindings=True)
		]

	##~~ SimpleApiPlugin mixin

	def turn_on(self, plugip, plugidx, username="", password=""):
		if self._settings.get(['singleRelay']):
			plugidx = ''
		self._domoticz_logger.debug("Turning on %s index %s." % (plugip, plugidx))
		plug = self.plug_search(self._settings.get(["arrSmartplugs"]),"ip",plugip,"idx",plugidx)
		try:
			strURL = "http://" + plugip + "/json.htm?type=command&param=switchlight&idx=" + str(plugidx) + "&switchcmd=On"
			if username != "":
				strURL = strURL + "&username=" + base64.b64encode(bytes(username)) + "&password=" + base64.b64encode(bytes(password))
			webresponse = requests.get(strURL)
			response = webresponse.json()
			chk = response["status"]
		except:
			self._domoticz_logger.error('Invalid ip or unknown error connecting to %s.' % plugip, exc_info=True)
			response = "Unknown error turning on %s index %s." % (plugip, plugidx)
			chk = "UNKNOWN"

		self._domoticz_logger.debug("Response: %s" % response)
		if chk == "OK":
			if plug["autoConnect"] and self._printer.is_closed_or_error():
				c = threading.Timer(int(plug["autoConnectDelay"]),self._printer.connect)
				c.start()
			if plug["sysCmdOn"]:
				t = threading.Timer(int(plug["sysCmdOnDelay"]),os.system,args=[plug["sysRunCmdOn"]])
				t.start()
			self._plugin_manager.send_plugin_message(self._identifier, dict(currentState="on",ip=plugip,idx=plugidx))
		else:
			self._domoticz_logger.debug(response)
			self._plugin_manager.send_plugin_message(self._identifier, dict(currentState="unknown",ip=plugip,idx=plugidx))

	def turn_off(self, plugip, plugidx, username="", password=""):
		self._domoticz_logger.debug("Turning off %s index %s." % (plugip, plugidx))
		plug = self.plug_search(self._settings.get(["arrSmartplugs"]),"ip",plugip,"idx",plugidx)
		try:
			if plug["sysCmdOff"]:
				self._domoticz_logger.debug("Running system command: %s in %s" % (plug["sysRunCmdOff"],plug["sysCmdOffDelay"]))
				t = threading.Timer(int(plug["sysCmdOffDelay"]),os.system,args=[plug["sysRunCmdOff"]])
				t.start()

			if plug["autoDisconnect"]:
				self._domoticz_logger.debug("Disconnnecting from printer")
				self._printer.disconnect()
				time.sleep(int(plug["autoDisconnectDelay"]))
			strURL = "http://" + plugip + "/json.htm?type=command&param=switchlight&idx=" + str(plugidx) + "&switchcmd=Off"
			if username != "":
				strURL = strURL + "&username=" + base64.b64encode(bytes(username)) + "&password=" + base64.b64encode(bytes(password))
			webresponse = requests.get(strURL)
			response = webresponse.json()
			chk = response["status"]
		except:
			self._domoticz_logger.error('Invalid ip or unknown error connecting to %s.' % plugip, exc_info=True)
			response = "Unknown error turning off %s index %s." % (plugip, plugidx)
			chk = "UNKNOWN"

		self._domoticz_logger.debug("Response: %s" % response)
		if chk == "OK":
			self._plugin_manager.send_plugin_message(self._identifier, dict(currentState="off",ip=plugip,idx=plugidx))
		else:
			self._domoticz_logger.debug(response)
			self._plugin_manager.send_plugin_message(self._identifier, dict(currentState="unknown",ip=plugip,idx=plugidx))

	def gcode_turn_off(self, plug):
		if plug["warnPrinting"] and self._printer.is_printing():
			self._domoticz_logger.debug("Not powering off %s since new print has started." % plug["label"])
		else:
			self.turn_off(plug["ip"],plug["idx"],plug["username"],plug["password"])

	def check_status(self, plugip, plugidx, username="", password=""):
		self._domoticz_logger.debug("Checking status of %s index %s." % (plugip, plugidx))
		if plugip != "":
			try:
				strURL = "http://" + plugip + "/json.htm?type=devices&rid=" + str(plugidx)
				if username != "":
					strURL = strURL + "&username=" + base64.b64encode(bytes(username)) + "&password=" + base64.b64encode(bytes(password))
				webresponse = requests.get(strURL)
				self._domoticz_logger.debug("%s index %s response: %s" % (plugip, plugidx, webresponse))
				response = webresponse.json()
				chk = response["result"][0]["Status"]
			except:
				self._domoticz_logger.error('Invalid ip or unknown error connecting to %s.' % plugip, exc_info=True)
				response = "unknown error with %s." % plugip
				chk = "UNKNOWN"

			self._domoticz_logger.debug("%s index %s is %s" % (plugip, plugidx, chk))
			if chk == "On":
				self._plugin_manager.send_plugin_message(self._identifier, dict(currentState="on",ip=plugip,idx=plugidx))
			elif chk == "Off":
				self._plugin_manager.send_plugin_message(self._identifier, dict(currentState="off",ip=plugip,idx=plugidx))
			else:
				self._domoticz_logger.debug(response)
				self._plugin_manager.send_plugin_message(self._identifier, dict(currentState="unknown",ip=plugip,idx=plugidx))

	def get_api_commands(self):
		return dict(turnOn=["ip","idx"],turnOff=["ip","idx"],checkStatus=["ip","idx"],connectPrinter=[],disconnectPrinter=[],sysCommand=["cmd"])

	def on_api_command(self, command, data):
		self._domoticz_logger.debug(data)
		if not user_permission.can():
			from flask import make_response
			return make_response("Insufficient rights", 403)
        
		if command == 'turnOn':
			if "username" in data and data["username"] != "":
				self._domoticz_logger.debug("Using authentication for %s." % "{ip}".format(**data))
				self.turn_on("{ip}".format(**data),"{idx}".format(**data),username="{username}".format(**data),password="{password}".format(**data))
			else:
				self.turn_on("{ip}".format(**data),"{idx}".format(**data))
		elif command == 'turnOff':
			if "username" in data and data["username"] != "":
				self._domoticz_logger.debug("Using authentication for %s." % "{ip}".format(**data))
				self.turn_off("{ip}".format(**data),"{idx}".format(**data),username="{username}".format(**data),password="{password}".format(**data))
			else:
				self.turn_off("{ip}".format(**data),"{idx}".format(**data))
		elif command == 'checkStatus':
			if "username" in data and data["username"] != "":
				self._domoticz_logger.debug("Using authentication for %s." % "{ip}".format(**data))
				self.check_status("{ip}".format(**data),"{idx}".format(**data),username="{username}".format(**data),password="{password}".format(**data))
			else:
				self.check_status("{ip}".format(**data),"{idx}".format(**data))
		elif command == 'connectPrinter':
			self._domoticz_logger.debug("Connecting printer.")
			self._printer.connect()
		elif command == 'disconnectPrinter':
			self._domoticz_logger.debug("Disconnecting printer.")
			self._printer.disconnect()
		elif command == 'sysCommand':
			self._domoticz_logger.debug("Running system command %s." % "{cmd}".format(**data))
			os.system("{cmd}".format(**data))

	##~~ Gcode processing hook

	def processGCODE(self, comm_instance, phase, cmd, cmd_type, gcode, *args, **kwargs):
		if gcode:
			if cmd.startswith("M8") and cmd.count(" ") >= 2:
				plugip = cmd.split()[1]
				plugidx = cmd.split()[2]
				for plug in self._settings.get(["arrSmartplugs"]):
					if plug["ip"].upper() == plugip.upper() and plug["idx"] == plugidx and plug["gcodeEnabled"]:
						if cmd.startswith("M80"):
							t = threading.Timer(int(plug["gcodeOnDelay"]),self.turn_on, [plug["ip"],plug["idx"]],{'username': plug["username"],'password': plug["password"]})
							t.start()
							self._domoticz_logger.debug("Received M80 command, attempting power on of %s index %s." % (plugip,plugidx))
							return
						elif cmd.startswith("M81"):
							t = threading.Timer(int(plug["gcodeOffDelay"]),self.gcode_turn_off, [plug])
							t.start()
							self._domoticz_logger.debug("Received M81 command, attempting power off of %s index %s." % (plugip,plugidx))
							return
						else:
							return
		elif cmd.startswith("@DOMOTICZ") and cmd.count(" ") == 1:
			plugidx = cmd.split()[1]
			for plug in self._settings.get(["arrSmartplugs"]):
				if plug["idx"] == plugidx and plug["gcodeEnabled"]:
					if cmd.startswith("@DOMOTICZON"):
						t = threading.Timer(int(plug["gcodeOnDelay"]),self.turn_on, [plug["ip"],plug["idx"]],{'username': plug["username"],'password': plug["password"]})
						t.start()
						self._domoticz_logger.debug("Received @DOMOTICZON command, attempting power on of %s index %s." % (plug["ip"],plugidx))
						return
					elif cmd.startswith("@DOMOTICZOFF"):
						t = threading.Timer(int(plug["gcodeOffDelay"]),self.gcode_turn_off, [plug])
						t.start()
						self._domoticz_logger.debug("Received @DOMOTICZOFF command, attempting power off of %s index %s." % (plug["ip"],plugidx))
						return
					else:
						return
			return

	##~~ Utility functions

	def lookup(self, dic, key, *keys):
		if keys:
			return self.lookup(dic.get(key, {}), *keys)
		return dic.get(key)

	def plug_search(self, list, key1, value1, key2, value2): 
		for item in list: 
			if item[key1] == value1 and item[key2] == value2: 
				return item

	##~~ Softwareupdate hook

	def get_update_information(self):
		return dict(
			domoticz=dict(
				displayName="OctoPrint-Domoticz",
				displayVersion=self._plugin_version,

				# version check: github repository
				type="github_release",
				user="jneilliii",
				repo="OctoPrint-Domoticz",
				current=self._plugin_version,

				# update method: pip
				pip="https://github.com/jneilliii/OctoPrint-Domoticz/archive/{target_version}.zip"
			)
		)

__plugin_name__ = "Domoticz"
__plugin_pythoncompat__ = ">=2.7,<4"

def __plugin_load__():
	global __plugin_implementation__
	__plugin_implementation__ = domoticzPlugin()

	global __plugin_hooks__
	__plugin_hooks__ = {
		"octoprint.comm.protocol.gcode.queuing": __plugin_implementation__.processGCODE,
		"octoprint.plugin.softwareupdate.check_config": __plugin_implementation__.get_update_information
	}

